### ISSUE TITLE
Invisible mobs/players

### ISSUE THREAD

**Issue by Yorkforce** [SpongePowered/SpongeForge#2764]
**I am currently running**

- SpongeForge version: 3712
- Forge version: 2831
- Java version: 8
- Operating System: Debian 9

- Plugins:
Minecraft, Minecraft Coder Pack, SpongeAPI, Sponge, Aliases, AntiWDL, AurionsVoteListener, AutoBroadcaster, Discord Bridge, EconomyLite, Enjin Minecraft Plugin, GriefPrevention, LuckPerms, MMCRestrict, MultiChatSponge, NuVotifier, Nucleus, Player Glow, PlayerShopsRPG, PwnFilter, RankUpper, SkyClaims

- Mods:
Minecraft, Minecraft Coder Pack, Forge Mod Loader, Minecraft Forge, SpongeAPI, Sponge, AE2 Stuff, Advancement Book, Animal Crops, Apotheosis, AppleCore, AppleSkin, Applied Energistics 2, Aroma1997Core, B.A.S.E, BD Lib, BNBGamingLib, Barrels, Drums, Storage & More, Baubles, Bed Patch, BiblioCraft, BnBGamingCore, Bonsai Trees, Bookshelf, Building Gadgets, Caliper, Calm Down Zombie Guy, Ceramics, Chargers, ChineseWorkshop, Chisel, Chisels & Bits, Click Machine, Clumps, CoFH Core, CoFH World, CodeChicken Lib, ComPatchedStorage, CommonCapabilities, Compact Machines 3, CompactStorage, Construct's Armory, ContentTweaker, Cooking for Blockheads, Corail Tombstone, CraftTweaker JEI Support, CraftTweaker2, Cucumber Library, Culinary Construct, Cyclic, Cyclops Core, Dark Utilities, Deep Mob Learning, Dimension Stages, DragonMurder, Elevator Mod, Ender Crop, Ender Utilities, EnderCore, EnderStorage, Extended Crafting, Extra Cells 2, Farming for Blockheads, FastFurnace, FastWorkbench, Fence Overhaul, FindMe, Flux Networks, FoamFix, FoamFixCore, Forge Microblocks, Forge Multipart CBE, Forgiving Void, Game Stages, Guide-API, Headcrumbs, Horse Power, Hunting Dimension, HydroGel, InControl, Industrial Foregoing, Inspirations, Integrated Dynamics, Integrated Tunnels, IntegratedDynamics-Compat, IntegratedTunnels-Compat, Iron Jetpacks, Item Stages, JourneyMap, Just Enough Items, KleeSlabs, LagGoggles, Limitless Structure Blocks, MCMultiPart, MTLib, MalisisCore, MalisisDoors, Mantle, MatterOverdrive: Legacy Edition, McJtyLib, Mekanism, Mekanism: Generators, Mercurius, Minecraft Multipart Plugin, Mob Stages, Mod Tweaker, ModPack Basic Tools, ModPack Utilities, More Buckets, More Cauldrons, Morpheus, Mystcraft, Mystical Agradditions, Mystical Agriculture, NuclearCraft, OG Dragon+, OpenComputers, OpenComputers (Core), OpenComputers Xnet Driver, OreExcavation, Parabox, Pickle Tweaks, PigUtils, Pipe Master 2000, Placebo, PlaneFix, PortalGun, Practical Logistics 2, Prestige, Primitive Crafting, ProjectE, RFToolsPower, Ranged Pumps, Real Filing Cabinet, RecipeStages, Redstone Flux, Resource Hogs, Rustic, Shadowfacts' Forgelin, Simple Generators, Simple Inventory sorting, Sky Grid, SkyBonsais, Slab Machines, SlimyBoyos, Smooth Font Core, Snad, SonarCore, SpongeForge, Squeezer Patch, Statues mod, Stuff A Sock In It, Super Sound Muffler, Surge, SwingThroughGrass, Sync, TallGates, TelePastries, Tesla Core Lib, Tesla Core Lib Registries, The Lost Cities, The Twilight Forest, The Weirding Gadget, Thermal Dynamics, Thermal Foundation, Thermal Innovation, Tinkers Tool Leveling, Tinkers' Complement, Tinkers' Construct, Tiny Progressions, Topography, TorchMaster, Translocators, Tree Growing Simulator 2016, Triumph, Twitchcrumbs, Universal Modifiers, Valkyrie Lib, ViesCraft, Waddles, Waila, Waila Stages, Wall-Jump!, What Are We Looking At, Wither Crumbs, World Utils, WorldEdit, XC Patch, XL Food Mod, XNet, YNot, Yoyos, ZenStages, bouncy_creepers, mob_grinding_utils, sky_orchards, uppers

**Issue Description**

This is a tough one to narrow down, we have had this issue on Sky Factory 4 since the modpack came out, when we were first testing the servers we noticed players on different Y levels were randomly going invisible. Then we started getting reports that friendlies were also going invisible. We have no idea what causes it just that it doesn't occur in single player nor has there been any reports on the Sky Factory 4 issue tracker.

Thankfully one of our players managed to record what was happening:
https://www.screencast.com/t/u8yKVDHtL

I don't know if this is spongeforge or not, but if it is and we could find a fix for this it would make a huge difference on our servers. Or if any other server owners are experiencing this, it could perhaps be down to a plugin or a mod?

One way we have been able to see it more often is via the animal crops mod, i had a normal spawning vanilla animal farm which seems to be completely fine, but when i used animal crop seeds to 'grow' some animals those would almost go invisible within a few seconds.
However, we have also seen it occur with players with no animal crops in sight.

There are no errors in the logs relating to this, it just appears to be totally random in most cases, most times the invisible mobs/players will reappear but it recurs frequently enough to see many comments in server chat regarding it.

**Comment by johnfriedrich** [comment]
Tbh not really sure if this tracker is the right place for it. I looked for this issue on the SF4 tracker and Discord aswell, but couldnt find it.
We have the same or similiar issue. For us every entity can be affected. For example lootcrate holograms from the lootcrate plugin at spawn, NPCs by the NPC plugin or holograms from our Holographic Displays bukkit port. But also vanilla friendly entities are going invisible, at least from player reports. But the modpack is full of mods and the issue appears fully randomly for us so I had no idea where to start debugging it. We have days with all spawn holograms surviving, sometimes they are missing, sometimes just the NPC, sometimes  everything is missing after 10 minutes.
SF 2713
Forge 2831

**Comment by SSKirillSS** [comment]
![image](https://user-images.githubusercontent.com/38190325/58370647-a4f1ea80-7f11-11e9-9fba-d55052c439b8.png)
Similar situation. Players are often not visible, and sometimes copies are spawning.

**Comment by phit** [comment]
apparently this can be caused by [Phosphor](https://minecraft.curseforge.com/projects/phosphor), are you all using that?

**Comment by Yorkforce** [comment]
That mod isn't in Sky Factory 4 @phit so I think we can rule it out.

**Comment by Shybella** [comment]
This can be replicated by placing a Feral Lamp from Torch Master. You need to let the Feral Lamp propagate light then remove it. It seems to really start glitching out bad when light sources are being removed.

Stoneblock2 and SK4 have TorchMaster.

**Comment by Yorkforce** [comment]
We saw it occur with invisible players on a fresh server we were testing before we released the SF4 servers, there were only 2 admins online and a fresh island with minimal items, there was definitely no feral lamps placed in the world.

It occured when one player was at a different y coordinate than the other, maybe slabs has something to do with it? There was one player island I visited and it seemed almost reproducible depending on where i stood, and it had a ton of slabs placed at different y coordinates too.

One thing we haven't seen is duplicate players, so not sure if SSKirillSS issue the same.

**Comment by TheFlash787** [comment]
We can confirm that a similar issue to this is happening on Omnifactory too, resulting in invisible mobs/players, too. The only difference to this is that many events are also not syncing with the client, resulting in a server-wide desync until the next server restart. For example, players may become invisible, but mobs will also stop moving and will freeze.

**Comment by TheFlash787** [comment]
I can also confirm that without TorchMaster, this can still occur.

**Comment by Yorkforce** [comment]
Thats interesting on the client desync, on SF4 we have have seen a few issues with clients not syncing properly. While it may not be related (although it doesnt seem to affect any other modpack we use) our hub server has alot of problems with SF4 client.

If a player has been on an SF4 server and they return to the hub (which is a bog standard 1.10 vanilla server) they can experience night time or rain, both of which are disabled on the hub, also they can end up with half the chunks not loading, leaving them suspended mid air unable to move or getting kicked for flying. 

Makes me wonder if one of the SF4 mods specifically separate client from server in some way and starts running its own 'timer'. If the same thing is occurring between 2 clients that could explain why other players and mobs are going invisible.

**Comment by clienthax** [comment]
Hm, im getting this and i have phosper in use.

**Comment by PhoenixTheSage** [comment]
@clienthax Same, with the latest 2890...going to remove Phosphor for a while and see if it still happens. If it does, I'm rolling back to 2825 and try again. Very strange.

**Comment by Zidane** [comment]
Can you folks try the latest SF again? This...may be fixed?

**Comment by PhoenixTheSage** [comment]
Will do, as it seems it's not Phosphor (on client nor server), thanks

Edit: Nope, still getting invisible mobs. 
Sometimes they spawn vanished, sometimes it happens on the fly. 

SpongeForge 3900 and Forge 1.12.2 - 14.23.5.2838
FTB Revelation 3.1.0
With and without Phosphor.

**Comment by PhoenixTheSage** [comment]
It's happening much more often on 3900 than 3890 actually. First time really noticing players going invisible too. Lots of entities frozen, while others normal. This seems to be a pretty critical bug to be hanging around since May, if you ask me xD

**Comment by kencinder** [comment]
I'm having this issue as well, currently running 3902. It does appear to be a sync issue with the client, as you can teleport away from somewhere that has entities appearing invisible, and teleport back and they appear briefly, then go invisible again.
I first noticed it with mob farm spawners, they'd spawn invisible, then noticed it with persistent sheep and cows.
No Phosphor here and took Optifine off the client to be certain it wasn't that. Can provide a mod list if need be.

**Comment by HrKure** [comment]
I have sf4 server and yes similar issue but it starts to happen more often when theres smeltrys loaded with liquids inside of them so it can be something to do with liquids in smeltrys.

**Comment by clienthax** [comment]
I don't think smelterys have anything todo with it, i get this in my mining area a lot.

**Comment by LXGaming** [comment]
Posting in this issue but there are 3 issues in total (so far) that are similar so this may or may not apply to any or all of the issues.

So to debug this issue I added some debug messages into a pre-existing Mixin [NetHandlerPlayServerMixin](https://github.com/SpongePowered/SpongeCommon/blob/8668a44baae03a86efc0111fa72bbf238ad41358/src/main/java/org/spongepowered/common/mixin/core/network/NetHandlerPlayServerMixin.java#L217) which was very convenient. This allowed me to see what packets the server is sending out.

After getting an armor stand to disappear on a clean setup of SkyFactory 4 with a custom Sponge build I check over the packets and noticed this: (Method names have been manually converted to mcp mappings)
```
[Server thread/INFO] [LX]: NetHandlerPlayServerMixin - Sending ChunkData Packet
java.lang.Exception: null
        at net.minecraft.network.NetHandlerPlayServer.redirect$impl$onSendPacket$zin000(NetHandlerPlayServer.java:3043) [pa.class:?]
        at net.minecraft.network.NetHandlerPlayServer.sendPacket(NetHandlerPlayServer.java:882) [pa.class:?]
        at net.minecraft.server.management.PlayerChunkMapEntry.sendPacket(PlayerChunkMapEntry.java:227) [ot.class:?]
        at net.minecraft.server.management.PlayerChunkMapEntry.handler$impl$UpdateBimoes$zjl000(PlayerChunkMapEntry.java:1064) [ot.class:?]
        at net.minecraft.server.management.PlayerChunkMapEntry.update(PlayerChunkMapEntry.java) [ot.class:?]
        at net.minecraft.server.management.PlayerChunkMap.tick(SourceFile:115) [ou.class:?]
        at net.minecraft.world.WorldServer.tick(WorldServer.java:227) [oo.class:?]
        at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:756) [MinecraftServer.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.updateTimeLightAndEntities(DedicatedServer.java:397) [nz.class:?]
        at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:668) [MinecraftServer.class:?]
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526) [MinecraftServer.class:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
```

Looking at [PlayerChunkMapEntryMixin](https://github.com/SpongePowered/SpongeCommon/blob/stable-7/src/main/java/org/spongepowered/common/mixin/core/server/management/PlayerChunkMapEntryMixin.java#L61) now aside from the method being spelt incorrectly (please fix). The issue here funnily enough has actually been documented in Sponge's code:
https://github.com/SpongePowered/SpongeCommon/blob/8668a44baae03a86efc0111fa72bbf238ad41358/src/main/java/org/spongepowered/common/mixin/api/mcp/world/WorldServerMixin_API.java#L278-L283

TL;DR "fullChunk packets cause the client to completely overwrite its current chunk with a new chunk instance. This causes weird issues, such as making any entities in that chunk invisible (until they leave it for a new chunk)"  - Aaron1011

**Comment by ImMorpheus** [comment]
PR merged.

Thank you @LXGaming

**Comment by Shybella** [comment]
This does seem to be resolved now. Players no one disappear.

Although, if you use a plugin such as GriefDefender/Griefprevention, things
like ambershop signs will disappear and reappear only inside a claim. This
could include other things inside the claim too, that's just from my
observations.
